properties/test_index_manipulation.py

```diff
@@ -0,0 +1,159 @@
+    )
+
+
+random.seed(123456)
+
+# Share to ensure we get unique names on each draw?
+UNIQUE_NAME = unique(strategy=xrst.names())
+DIM_NAME = xrst.dimension_names(name_strategy=UNIQUE_NAME, min_dims=1, max_dims=1)
+
+
+class DatasetStateMachine(RuleBasedStateMachine):
+    indexed_dims = Bundle("indexed_dims")
+    multi_indexed_dims = Bundle("multi_indexed_dims")
+
+    def __init__(self):
+        super().__init__()
+        self.dataset = Dataset()
+        self.check_default_indexes = True
+
+    @rule(var=xrst.index_variables(dims=DIM_NAME), target=indexed_dims)
+    def add_dim_coord(self, var):
+        (name,) = var.dims
+        # dim coord
+        self.dataset[name] = var
+        # non-dim coord of same size; this allows renaming
+        self.dataset[name + "_"] = var
+        return name
+
+    @rule(dim=st.one_of(consumes(indexed_dims), consumes(multi_indexed_dims)))
+    def reset_index(self, dim):
+        self.check_default_indexes = False
+        note(f"> resetting {dim}")
+        self.dataset = self.dataset.reset_index(dim)
+
+    @rule(
+        newname=UNIQUE_NAME,
+        oldnames=st.lists(consumes(indexed_dims), min_size=1),
+        target=multi_indexed_dims,
+    )
+    def stack(self, newname, oldnames):
+        note(f"> stacking {oldnames} as {newname}")
+        self.dataset = self.dataset.stack({newname: oldnames})
+        return newname
+
+    # TODO: add st.none() to dim
+    @rule(dim=consumes(multi_indexed_dims), target=indexed_dims)
+    def unstack(self, dim):
+        if dim is not None:
+            pd_index = self.dataset.xindexes[dim].index
+            assume(pd_index.is_unique)
+        note(f"> unstacking {dim}")
+        self.dataset = self.dataset.unstack(dim)
+        if dim is not None:
+            return multiple(pd_index.names)
+        else:
+            # TODO Fix this when adding st.none()
+            return multiple()
+
+    @rule(newname=UNIQUE_NAME, oldname=consumes(indexed_dims))
+    def rename_vars(self, newname, oldname):
+        # benbovy: "skip the default indexes invariant test when the name of an
+        # existing dimension coordinate is passed as input kwarg or dict key
+        # to .rename_vars()."
+        self.check_default_indexes = False
+        note(f"> renaming {oldname} to {newname}")
+        self.dataset = self.dataset.rename_vars({oldname: newname})
+
+    @rule(data=st.data(), dim=consumes(indexed_dims), target=indexed_dims)
+    @precondition(lambda self: len(self.dataset._variables) >= 2)
+    def swap_dims(self, data, dim):
+        ds = self.dataset
+        choices = [name for name, var in ds._variables.items() if var.dims == (dim,)]
+        # TODO: is there a better way to skip if choices == []
+        # note(choices)
+        # if not choices:
+        #     return dim
+        # Can only swap to a variable with the same dim
+        to = data.draw(st.sampled_from(choices))
```

Comment:
@Zac-HD quick q: I need to skip tests when choices is empty but using	`filter` or `assume` tells me that this is filtering too much. Is there some other approach, or shall I just disable the filter_too_much check?

Reply:
I think ignoring it would lead to very slow tests, so let's try to make this fast.

The obvious way to do _that_ is to have a precondition that there exist (dimensions with) valid choices among the variables, and then to arrange your `dims` strategy to choose such a dimension.  

Unfortunately there's currently no way to put a filter on a `consumes(bundle)` strategy, which seems something of an oversight - it'd just take a few dozen lines of code.  Until someone does that, I'd probably just turn your `indexed_dims` bundle into an instance variable - it's also usable in a precondition that way, and you can `draw(st.sampled_from(self.indexed_dims))` inside the test easily enough.